src/bin.ts

```diff
@@ -180,52 +187,57 @@ const deploy = async ({ yes, bucket }: { yes: boolean, bucket: string }) => {
-                const promise = new S3.ManagedUpload({
-                    service: s3,
-                    params: {
-                        Bucket: config.bucketName,
-                        Key: key,
-                        Body: fs.createReadStream(path),
-                        ACL: config.acl === null ? undefined : (config.acl || 'public-read'),
-                        ContentType: mime.getType(path) || 'application/octet-stream',
-                        ...getParams(key, params)
+            uploadQueue.push(asyncify(async () => {
+                const key = createSafeS3Key(relative(publicDir, path));
+                const stream = fs.createReadStream(path);
+                const hashStream = stream.pipe(createHash('md5').setEncoding('hex'));
+                const data = await streamToPromise(hashStream)
+    
+                const tag = `"${data}"`;
+                const object = objects.find(object => object.Key === key && object.ETag === tag);
+    
+                isKeyInUse[key] = true;
+            
+                if (!object) {
+                    try {
+                        const upload = new S3.ManagedUpload({
+                            service: s3,
+                            params: {
+                                Bucket: config.bucketName,
+                                Key: key,
+                                Body: fs.createReadStream(path),
+                                ACL: config.acl === null ? undefined : (config.acl || 'public-read'),
+                                ContentType: mime.getType(path) || 'application/octet-stream',
+                                ...getParams(key, params)
+                            }
+                        })
+        
+                        upload.on('httpUploadProgress', (evt) => {
+                            spinner.text = chalk`Syncing...\n{dim   Uploading {cyan ${key}} ${evt.loaded.toString()}/${evt.total.toString()}}`
+                        })
+        
+                        await upload.promise();
+                        spinner.text = chalk`Syncing...\n{dim   Uploaded {cyan ${key}}}`;
+        
+                    } catch (ex) {
+                        console.error(ex);
+                        process.exit(1);
                     }
-                }).promise();
-                promises.push(promise);
-                await promise;
-                spinner.text = chalk`Syncing...\n{dim   Uploaded {cyan ${key}}}`;
-            } catch (ex) {
-                spinner.fail(chalk`Upload failure for object {cyan ${key}}`);
-                console.error(pe.render(ex));
-                process.exit(1);
-            }
+                }
+
+            }));
         });
 
         // now we play the waiting game.
         await streamToPromise(stream as any as Readable); // todo: find out why the typing won't allow this as-is
-        await Promise.all(promises);
-
+        await promisifiedParallelLimit(uploadQueue, 20);
```

Comment:
This limit should be made a config option. As the deployment breaks on large sites uploading with slow internet.